src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -177,23 +177,45 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
     }
 
     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Fuction redirect to random page
+        // Fuction redirect to random page of the same type
         // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
         Long mealId = Long.parseLong(getParameter(request, "id", "0"));
+        String mealType = null;
 
         Query query = new Query("Meal");
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
+
+        for (Entity entity : results.asIterable()) {
+            Long id = (Long)entity.getProperty("id");
+            if (id != null && id.equals(mealId)) {
+                mealType = (String)entity.getProperty("type");
+            }
+        }
+
+        ArrayList<Long> sameTypeIdList = new ArrayList<>();
         ArrayList<Long> idList = new ArrayList<>();
         for (Entity entity : results.asIterable()) {
             Long id = (Long)entity.getProperty("id");
+            String type = (String)entity.getProperty("type");
             if (id != null && !id.equals(mealId)) {
                 idList.add(id);
             }
+            if (id != null && !id.equals(mealId) && type != null && type.equals(mealType)) {
+                sameTypeIdList.add(id);
+            }
         }
+
+
         Random rand = new Random(); 
         int index = rand.nextInt(idList.size());
         Long randomId = idList.get(index);
```

Comment:
Request the entire list of entries from the database to take a random one only *after* it turns out there's no truly similar meals found.